word_language_model/model.py

```diff
@@ -1,57 +1,165 @@
+import math
+import torch
 import torch.nn as nn
+import torch.nn.functional as F
 
-class RNNModel(nn.Module):
-    """Container module with an encoder, a recurrent module, and a decoder."""
+class Seq2SeqModel(nn.Module):
+    """Container module with an encoder, a recurrent or transformer module, and a decoder."""
 
-    def __init__(self, rnn_type, ntoken, ninp, nhid, nlayers, dropout=0.5, tie_weights=False):
-        super(RNNModel, self).__init__()
+    def __init__(self, model_type, ntoken, ninp, nhid, nlayers, dropout=0.5, tie_weights=False, nhead=None):
+        super(Seq2SeqModel, self).__init__()
         self.drop = nn.Dropout(dropout)
         self.encoder = nn.Embedding(ntoken, ninp)
-        if rnn_type in ['LSTM', 'GRU']:
-            self.rnn = getattr(nn, rnn_type)(ninp, nhid, nlayers, dropout=dropout)
+        self.model_type = model_type
+        self.nhead = nhead
+        self.ninp = ninp
+
+        if self.model_type in ['LSTM', 'GRU']:
+            self.model = getattr(nn, self.model_type)(ninp, nhid, nlayers, dropout=dropout)
+        elif self.model_type == 'Transformer':
+            try:
+                from torch.nn import TransformerEncoder, TransformerEncoderLayer
+            except:
+                raise ImportError('TransformerEncoder module does not exist in PyTorch 1.1 or lower.')
+            if nhead is None:
+                assert self.model_type != 'Transformer'
+            self.src_mask=None
+            self.pos_encoder = PositionalEncoding(ninp, dropout)
+            encoder_layers = TransformerEncoderLayer(ninp, nhead, nhid, dropout)
+            self.model = TransformerEncoder(encoder_layers, nlayers)
         else:
             try:
-                nonlinearity = {'RNN_TANH': 'tanh', 'RNN_RELU': 'relu'}[rnn_type]
+                nonlinearity = {'RNN_TANH': 'tanh', 'RNN_RELU': 'relu'}[self.model_type]
             except KeyError:
                 raise ValueError( """An invalid option for `--model` was supplied,
                                  options are ['LSTM', 'GRU', 'RNN_TANH' or 'RNN_RELU']""")
-            self.rnn = nn.RNN(ninp, nhid, nlayers, nonlinearity=nonlinearity, dropout=dropout)
+            self.model = nn.RNN(ninp, nhid, nlayers, nonlinearity=nonlinearity, dropout=dropout)
         self.decoder = nn.Linear(nhid, ntoken)
 
-        # Optionally tie weights as in:
-        # "Using the Output Embedding to Improve Language Models" (Press & Wolf 2016)
-        # https://arxiv.org/abs/1608.05859
-        # and
-        # "Tying Word Vectors and Word Classifiers: A Loss Framework for Language Modeling" (Inan et al. 2016)
-        # https://arxiv.org/abs/1611.01462
-        if tie_weights:
-            if nhid != ninp:
-                raise ValueError('When using the tied flag, nhid must be equal to emsize')
-            self.decoder.weight = self.encoder.weight
+        if self.model_type == 'Transformer':
```

Comment:
It should be neater to branch within "init_weights" instead of having a separate function. You can assert that tie_weights isn't used together with transformer.